### PR TITLE
Improve config documentation

### PR DESCRIPTION
**Pull request by awav**

### CODE REVIEW

File: gpflow/config/__config__.py [hunk 1]
```diff
@@ -1,3 +1,40 @@
+"""
+This is a private module that manages GPflow configuration.
+
+The module provides functions to modify default settings of GPflow, such as:
+the standard float precision and integer type, the positive transformation
+type, a value for a minimum shift from zero for the positive transformation,
+an output format for the summary printing.
```
Comment (vdutor) [inline gpflow/config/__config__.py]: Nit:
```suggestion
- the standard float precision and integer type
- the type of positive transformation
- a value for a minimum shift from zero for the positive transformation
- an output format for `gpflow.utilities.print_summary`
```

File: gpflow/config/__config__.py [hunk 2]
```diff
@@ -1,3 +1,40 @@
...
+
+The module holds global configuration :class:`Config` variable that stores all
+setting values.
+
+The environment variables is an alternative way for changing default GPflow
```
Comment (vdutor) [inline gpflow/config/__config__.py]:
```suggestion
The environment variables is an alternative way for changing the default GPflow
```

File: gpflow/config/__config__.py [hunk 3]
```diff
@@ -1,3 +1,40 @@
...
+configuration.
+
+.. warning::
+    The user has to set environment variables before running python
+    interpreter to modify the configuration.
+
+Full set of environment variables and available options:
+    ```
+    GPFLOW_INT: "int16", "int32", or "int64"
```
Comment (alexisboukouvalas) [inline gpflow/config/__config__.py]: can these be any type from https://www.tensorflow.org/api_docs/python/tf/dtypes/DType ?

File: gpflow/config/__config__.py [hunk 4]
```diff
@@ -135,36 +172,50 @@ def config() -> Config:
...
     return config().positive_bijector
 
 
 def default_positive_minimum():
+    """Shift constant that GPflow adds to all positive constraints."""
     return config().positive_minimum
 
 
 def default_summary_fmt():
+    """Summary printing format."""
```
Comment (alexisboukouvalas) [inline gpflow/config/__config__.py]:
```suggestion
    """Summary printing format as understood by tabulate ."""
```

File: gpflow/config/__config__.py [hunk 5]
```diff
@@ -135,36 +172,50 @@ def config() -> Config:
...
 def default_float():
+    """Returns default float type"""
     return config().float
 
 
 def default_jitter():
+    """
+    The jitter is a constant that GPflow adds to diagonals of matrices
+    to achieve numerical stability of the system when conditional number of
+    matrices is large and therefore matrix is nearly singular.
```
Comment (vdutor) [inline gpflow/config/__config__.py]:
```suggestion
    The jitter is a constant that GPflow adds to the diagonal of matrices
    to achieve numerical stability of the system when the condition number 
    of the associated matrices is large, and therefore the matrices nearly singular.
```

File: gpflow/config/__config__.py [hunk 6]
```diff
@@ -1,3 +1,40 @@
...
+    GPFLOW_POSITIVE_MINIMUM: Any positive float number
+    GPFLOW_SUMMARY_FMT: "notebook" or any other format that
+        :mod:`tabulate` can handle.
+    GPFLOW_JITTER: Any positive float number
+    ```
+
+The user also can change GPflow configuration temporarily with a context
+manager :func:`as_context`:
+
+>>> config = Config(jitter=1e-5)
```
Comment (alexisboukouvalas) [inline gpflow/config/__config__.py]: I don't see any examples of this in the notebooks - should we add an example usage?

File: gpflow/config/__config__.py [hunk 7]
```diff
@@ -135,36 +172,50 @@ def config() -> Config:
...
-    """Update GPflow config"""
+    """Update GPflow config with new settings from `new_config`."""
     global __config
     __config = new_config
 
 
 def set_default_int(value_type):
+    """
+    Sets default integer type. Available options are np.int16, np.int32,
+    or np.int64.
```
Comment (vdutor) [inline gpflow/config/__config__.py]: Nit:
```suggestion
    Sets default integer type. Available options are `np.int16`, `np.int32`,
    or `np.int64`.
```

File: gpflow/config/__config__.py [hunk 8]
```diff
@@ -177,6 +228,10 @@ def set_default_int(value_type):
 
 
 def set_default_float(value_type):
+    """
+    Sets default float type. Available options are np.float16, np.float32,
+    or np.float64.
```
Comment (vdutor) [inline gpflow/config/__config__.py]: Nit:
```suggestion
    Sets default float type. Available options are `np.float16`, `np.float32`,
    or `np.float64`.
```

File: gpflow/config/__config__.py [hunk 9]
```diff
@@ -189,6 +244,12 @@ def set_default_float(value_type):
 
 
 def set_default_jitter(value: float):
+    """
+    Sets constant jitter value.
+    The jitter is a constant that GPflow adds to diagonals of matrices
+    to achieve numerical stability of the system when conditional number of
+    matrices is large and therefore matrix is nearly singular.
```
Comment (vdutor) [inline gpflow/config/__config__.py]:
```suggestion
    The jitter is a constant that GPflow adds to the diagonal of matrices
    to achieve numerical stability of the system when the condition number 
    of the associated matrices is large, and therefore the matrices nearly singular.
```

File: gpflow/config/__config__.py [hunk 10]
```diff
@@ -201,6 +262,10 @@ def set_default_jitter(value: float):
 
 
 def set_default_positive_bijector(value: str):
+    """
+    Sets positive bijector type.
+    There two options: "exp" and "softplus".
```
Comment (st--) [inline gpflow/config/__config__.py]:
```suggestion
    There are currently two options implemented: "exp" and "softplus".
```

File: gpflow/config/__config__.py (hunk 4 above)
Comment (st--) [reply gpflow/config/__config__.py]: there is also the "notebook" special-case

File: gpflow/config/__config__.py (hunk 6 above)
Comment (st--) [reply gpflow/config/__config__.py]: yes, would be great to add this to one of the intro notebooks (intro_to_gpflow2 maybe)?

File: gpflow/config/__config__.py (hunk 3 above)
Comment (awav) [reply gpflow/config/__config__.py]: It is a section about environment variable and you will pass a string there. Although, in a setter, you are allowed to pass TensorFlow and NumPy types.

File: gpflow/config/__config__.py (hunk 2 above)
Comment (st--) [reply gpflow/config/__config__.py]:
```suggestion
Environment variables are an alternative way for changing the default GPflow
```

File: gpflow/config/__config__.py [hunk 11]
```diff
@@ -1,3 +1,40 @@
...
+Full set of environment variables and available options:
+    ```
+    GPFLOW_INT: "int16", "int32", or "int64"
+    GPFLOW_FLOAT: "float16", "float32", or "float64"
+    GPFLOW_POSITIVE_BIJECTOR: "exp" or "softplus"
+    GPFLOW_POSITIVE_MINIMUM: Any positive float number
+    GPFLOW_SUMMARY_FMT: "notebook" or any other format that
+        :mod:`tabulate` can handle.
+    GPFLOW_JITTER: Any positive float number
+    ```
```
Comment (st--) [inline gpflow/config/__config__.py]: should these be wrapped in ``` ?

File: gpflow/config/__config__.py [hunk 12]
```diff
@@ -1,3 +1,40 @@
...
+
+The user also can change GPflow configuration temporarily with a context
```
Comment (st--) [inline gpflow/config/__config__.py]:
```suggestion
The user can also change the GPflow configuration temporarily with a context
```

File: gpflow/config/__config__.py (hunk 6 above)
Comment (awav) [reply gpflow/config/__config__.py]: I will address it in a separate PR.

**Pull request merged**